feat: revamp `particleAdapter`

TABASCOatw opened the pull request:
As an alternative to the previous PR, this purely updates the adapter to function with the new version of Particle's `solana-wallet` SDK, restoring standard functionality to the adapter (which it currently lacks due to the outdated structure and `solana-wallet` version). 

This removes the previous social login display customization changes & abides by the 1 SDK rule.

jordaaash commented on packages/wallets/particle/src/adapter.ts:
```diff
@@ -37,12 +45,37 @@ export class ParticleAdapter extends BaseMessageSignerWalletAdapter {
-    constructor(config: ParticleAdapterConfig = {}) {
+    private _particleNetwork: ParticleNetwork | null = null;
+
+    constructor(config: ParticleAdapterConfig) {
```
This is also a breaking change -- any app that uses `ParticleWalletAdapter` will get a type error because `config` is now a required argument.

Is there any way you can refactor this to make this not required?

TABASCOatw replied:
Sure, should be fixed now

jordaaash commented on packages/wallets/particle/src/adapter.ts:
```diff
@@ -74,17 +108,29 @@ export class ParticleAdapter extends BaseMessageSignerWalletAdapter {
-                wallet = new WalletClass(new ParticleClass(this._config?.config).auth);
+                particleNetwork = new ParticleClass(this._config?.config as Config);
+                if (!particleNetwork.auth.isLogin()) {
+                    await particleNetwork.auth.login(authOptions);
+                }
             } catch (error: any) {
                 throw new WalletConfigError(error?.message, error);
             }
 
+            this._particleNetwork = particleNetwork;
+
+            let wallet: SolanaWallet;
             try {
-                await wallet.connect();
+                wallet = new WalletClass(particleNetwork.auth);
             } catch (error: any) {
-                throw new WalletConnectionError(error?.message, error);
+                throw new WalletConfigError(error?.message, error);
```
Can you describe a bit what this new code is doing? Is the `wallet` instance always connected? What about the case where `!particleNetwork.auth.isLogin()` is `false`?

TABASCOatw replied:
If `isLogin` returns false, then it initiates login through `particleNetwork.auth.login`, otherwise if it returns true, an assumption can be made that the user already has an open instance with Particle, in which this account is either used instead or the connection is already initiated. The changes here to how `wallet` is handled is a response to a shift in the architecture of `solana-wallet`

jordaaash reviewed (changes requested):
Thanks, much easier to review. Still have some questions, primary about the configuration interface.

TABASCOatw commented on packages/wallets/particle/package.json:
```diff
@@ -35,7 +35,8 @@
     },
     "dependencies": {
-        "@particle-network/solana-wallet": "^0.5.6",
+        "@metamask/eth-sig-util": "^7.0.1",
```
During the previous review there were some errors thrown due to this missing (`@metamask/eth-sig-util`), including it as a direct dependency solves some issues within `solana-wallet` in this case; it's commonly presented as an engine error when this is missing or incompatible.

jordaaash replied:
Since this isn't a dependency we use here, it should be removed. If it's needed by your SDK, it should be added as a dependency there.

TABASCOatw commented on packages/wallets/particle/package.json:
```diff
@@ -1,6 +1,6 @@
 {
     "name": "@solana/wallet-adapter-particle",
-    "version": "0.1.10",
+    "version": "0.1.11",
```
Gotcha, sure I can revert this and open a new PR with an alternative branch; will review the other requested changes then make the swap

TABASCOatw commented on packages/wallets/particle/src/adapter.ts:
```diff
@@ -17,8 +16,17 @@ import {
 export interface ParticleAdapterConfig {
-    config?: Config;
+    config: NestedConfig;
+    preferredAuthType?: string;
```
Should be fixed now, made it optional

TABASCOatw commented on packages/wallets/particle/src/adapter.ts:
```diff
@@ -37,12 +45,37 @@ export class ParticleAdapter extends BaseMessageSignerWalletAdapter {
+        };
+    }
+
+    public get auth(): ParticleNetwork['auth'] {
+        if (!this._particleNetwork) {
+            throw new Error('ParticleNetwork is not initialized.');
+        }
+        return this._particleNetwork.auth;
```
This is to expose `.auth` on the instance of the adapter, which opens up a ton of functionality (manual logins, user info retrieval, disconnect, etc.) - `.auth.{}` is a sort of controller for the WaaS instance, thus exposing it on the instance itself allows it to be used accordingly.

jordaaash replied:
Okay, I think it might make more sense to expose the entire `particleNetwork` object (several other adapters expose this kind of thing as `adapter.wallet`).

And maybe rename the internal `_particleNetwork` as `_wallet`.

jordaaash replied:
This way the adapter is minimally aware of the wallet you want to expose, since it doesn't use it directly, and just exposes the entire thing. If you add or remove functionality later, we won't need to update the adapter code, just the SDK version.

TABASCOatw commented on packages/wallets/particle/src/adapter.ts:
```diff
@@ -74,17 +117,29 @@ export class ParticleAdapter extends BaseMessageSignerWalletAdapter {
-            let wallet: SolanaWallet;
+            let particleNetwork: ParticleNetwork;
+
+            const authOptions: any = {};
```
Unfortunately not directly without importing a second SDK that contains the specific type here. I can port the type directly although would require somewhat of a refactor. And sure, just updated it with lint

jordaaash replied:
Let's import that SDK for its type then. We try to avoid `any` unless absolutely necessary.

TABASCOatw commented:
GM @jordaaash! Anything else you'd like to be done here? Happy to make any other changes; let me know 👍

TABASCOatw commented:
Just checking in here again @jordaaash

TABASCOatw commented:
Hey @jordaaash! Happy New Year! Just wanted to bump this and see if there was anything else that I could do here to push it through?

jordaaash commented:
Catching up on this now. Sorry for the delay!

TABASCOatw commented:
> Catching up on this now. Sorry for the delay!

No problem at all, thank you!

jordaaash commented on packages/wallets/particle/src/adapter.ts:
```diff
@@ -37,12 +45,46 @@ export class ParticleAdapter extends BaseMessageSignerWalletAdapter {
+    }
+
+    public get particle(): ParticleNetwork {
+        if (!this._particleNetwork) {
+            throw new Error('ParticleNetwork is not initialized.');
+        }
+        return this._particleNetwork;
```
Can you refactor this so that the getter doesn't throw and just returns the value, which may be `null`?

jordaaash commented on packages/wallets/particle/src/adapter.ts:
```diff
@@ -74,17 +117,29 @@ export class ParticleAdapter extends BaseMessageSignerWalletAdapter {
                 throw new WalletConfigError(error?.message, error);
             }
 
+            this._particleNetwork = particleNetwork;
```
This should be done at the end, where we set `this._wallet`. This prevents any state changes from occurring if the initialization of anything fails in the `connect` function.

TABASCOatw replied:
Sure np, moved it down to where this._wallet is set.

jordaaash reviewed (changes requested):
Okay, a few small changes left. I enabled the build to run, but it looks broken. Can you commit the pnpm lockfile changes? I also ran the build locally, and it fails with this error:

```
@solana/wallet-adapter-particle:build: ../../../node_modules/.pnpm/@particle-network+solana-wallet@1.2.1_@particle-network+auth@0.5.6_@solana+web3.js@1.78.0_bs58@4.0.1/node_modules/@particle-network/solana-wallet/lib/types/solana-wallet.d.ts(1,16): error TS2305: Module '"@particle-network/auth"' has no exported member 'LoginOptions'.
```

TABASCOatw commented:
> Okay, a few small changes left. I enabled the build to run, but it looks broken. Can you commit the pnpm lockfile changes? I also ran the build locally, and it fails with this error:
> 
> ```
> @solana/wallet-adapter-particle:build: ../../../node_modules/.pnpm/@particle-network+solana-wallet@1.2.1_@particle-network+auth@0.5.6_@solana+web3.js@1.78.0_bs58@4.0.1/node_modules/@particle-network/solana-wallet/lib/types/solana-wallet.d.ts(1,16): error TS2305: Module '"@particle-network/auth"' has no exported member 'LoginOptions'.
> ```

Just pushed changes addressing those additional comments and fixed the build issue - I went ahead and added my pnpm lockfile to the commit to. The build is working locally for me, could you check again on your side?

Additionally, important to note that this change adds an additional dev dependency to the adapter package, `@metamask/eth-sig-util`, I tried a few workarounds but it isn't saved as a peer dep and instead needs to be included here to avoid that build issue you ran into, the other option would be to ignore it but could cause some issues.

jordaaash commented:
Followup PR to fix and release this @ https://github.com/solana-labs/wallet-adapter/pull/877

jordaaash commented:
Released via https://github.com/solana-labs/wallet-adapter/pull/878
